agent/src/main/java/org/openremote/agent/protocol/lorawan/chirpstack/ChirpStackProtocol.java

```diff
@@ -125,4 +297,177 @@ protected boolean configureMQTTWriteValueTemplate(Attribute<?> attribute, MQTTAg
         }
         return true;
     }
+
+    @Override
+    protected String generateAssetId(String devEui) {
+        return UniqueIdentifierGenerator.generateId("ChirpStack_" + getAgent().getId() + devEui);
+    }
+
+    private int getPort() {
+        return getAgent().getPort().orElse(isSecureGRPC() ? 443 : 80);
+    }
+
+    private boolean isSecureGRPC() {
+        return getAgent().getSecureGRPC().orElse(false);
+    }
+
+    private String getGRPCClientUri() {
+        return "chirpstack-grpc://" + getAgent().getHost().orElse("-") + ":" + getPort()
+                                    + "/?clientId=" + getAgent().getClientId().orElse("-");
+    }
+
+    private ManagedChannel createChannel(String host, int port, boolean isSecure) {
+        ManagedChannel channel = null;
+
+        if (isSecure) {
+            channel = ManagedChannelBuilder.forAddress(host, port)
+                .useTransportSecurity()
+                .build();
+        } else {
+            channel = ManagedChannelBuilder.forAddress(host, port)
+                .usePlaintext()
+                .build();
+        }
+
+        return channel;
+    }
+
+    private Optional<Asset<?>> createAsset(String devEUI, String assetTypeName, DeviceListItem deviceItem, DeviceProfile deviceProfile) {
+        if (isNullOrEmpty(devEUI) || isNullOrEmpty(assetTypeName) || deviceItem == null || deviceProfile == null) {
+            return Optional.empty();
+        }
+
+        CsvRecord record = new CsvRecord();
+        record.setDevEUI(devEUI);
+        record.setAssetTypeName(assetTypeName);
+        Optional.ofNullable(deviceItem.getName())
+            .filter(name -> !name.isEmpty())
+            .ifPresent(record::setName);
+
+        return Optional.ofNullable(assetTypeName)
```

Comment:
`'Optional.ofNullable()' with non-null argument should be replaced with 'Optional.of()'`